thirdparty/hairer/radau5_c_py.pyx

```diff
@@ -106,21 +97,146 @@ cdef int callback_solout(integer* nrsol, doublereal* xosol, doublereal* xsol, do
     cdef np.ndarray[double, ndim=1, mode="c"]y_py = np.empty(nsolu[0], dtype = np.double)
     cdef np.ndarray[double, ndim=1, mode="c"]cont_py = np.empty(4*nsolu[0], dtype = np.double)
     cdef np.ndarray[double, ndim=1, mode="c"]werr_py = np.empty(nsolu[0], dtype = np.double)
-    c2py(y_py, y, nsolu[0])
-    c2py(cont_py, cont, 4*nsolu[0])
-    c2py(werr_py, werr, nsolu[0])
+    c2py_d(y_py, y, nsolu[0])
+    c2py_d(cont_py, cont, 4*nsolu[0])
+    c2py_d(werr_py, werr, nsolu[0])
 
     irtrn[0] = (<object>solout_PY)(nrsol[0], xosol[0], xsol[0],
                                    y_py, cont_py, werr_py,
                                    lrc[0], irtrn[0])
 
     return irtrn[0]
 
+cdef class RadauSuperLUaux:
+    """Auxiliary data structure to have the memory ownership of internal
+    and auxiliary superLU data structures available on the highest possible level.
+    """
+    cdef SuperLU_aux_d* superLU_aux_struct_d # real
+    cdef SuperLU_aux_z* superLU_aux_struct_z # complex
+    ## data storage of sparse jacobian
+    cdef double* jac_data
+    cdef int* jac_indicies
+    cdef int* jac_indptr
+    
+    cpdef int initialize(self, int nprocs, int n, int nnz):
+        self.superLU_aux_struct_d = superlu_init_d(nprocs, n, nnz)
+        self.superLU_aux_struct_z = superlu_init_z(nprocs, n, nnz)
+
+        radau_sparse_aux_init(&self.jac_data, &self.jac_indicies, &self.jac_indptr, nnz, n)
+        return 0
+
+    cpdef int finalize(self):
+        radau_sparse_aux_finalize(&self.jac_data, &self.jac_indicies, &self.jac_indptr)
+        cdef int ret
+        ret = superlu_finalize_d(self.superLU_aux_struct_d)
+        if ret != 0:
+            return ret
+        ret = superlu_finalize_z(self.superLU_aux_struct_z)
+        return ret
+
+cdef int callback_jac_sparse(int n, double *x, double *y, int *nnz,
+                             double * data, int *indices, int *indptr,
+                             doublereal* rpar, integer* ipar,
+                             void* jac_PY):
+    """
+    Internal callback function to enable call to Python based evaluation of sparse (csc) jacobians.
+
+    Checks if jacobian includes the diagonal.
+    If not, it adds corresponding zeroes in the CSC format.
+    This is since the resulting system matrix will add values on the diagonal.
+    """
+    cdef np.ndarray[double, ndim=1, mode="c"]y_py = np.empty(n, dtype = np.double)
+    c2py_d(y_py, y, n)
+
+    J = (<object>jac_PY)(x[0], y_py)
+    if not isinstance(J, sps.csc.csc_matrix):
+        return -1
+
+    if J.nnz > nnz[0]:
+        return J.nnz
+
+    cdef np.ndarray[double, mode="c", ndim=1] jac_data_py = J.data.astype(np.double)
+    cdef np.ndarray[int, mode="c", ndim=1] jac_indices_py = J.indices.astype(np.intc)
+    cdef np.ndarray[int, mode="c", ndim=1] jac_indptr_py = J.indptr.astype(np.intc)
+
+    ## Check if any diagonal entries are missing in Jacobian
+
+    missing = [] ## list of diagonal entries that are missing

```

Comment:
I propose to move all the checking and adding of the diagonal down to C and hide this from the callback functions. The callback function should be unaware that we add it internally.

I also suggest that checking / adding of the diagonal is a separate method. See for instance SUNDIALS SparseAddIdentityMat method.

Reply:
Great!